hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java

```diff
@@ -360,20 +360,33 @@ private void watchForCommit(boolean bufferFull) throws IOException {
   }
 
   private CompletableFuture<ContainerProtos.
-      ContainerCommandResponseProto> executePutBlock()
-      throws IOException {
+      ContainerCommandResponseProto> executePutBlock(boolean close,
+      boolean force) throws IOException {
     checkOpen();
     long flushPos = totalDataFlushedLength;
-    Preconditions.checkNotNull(bufferList);
-    final List<ChunkBuffer> byteBufferList = bufferList;
-    bufferList = null;
-    Preconditions.checkNotNull(byteBufferList);
+    final List<ChunkBuffer> byteBufferList;
+    if (!force) {
+      Preconditions.checkNotNull(bufferList);
+      byteBufferList = bufferList;
+      bufferList = null;
+      Preconditions.checkNotNull(byteBufferList);
+    } else {
+      byteBufferList = null;
+    }
 
     CompletableFuture<ContainerProtos.
         ContainerCommandResponseProto> flushFuture;
     try {
+      BlockData blockData = containerBlockData.build();
+      if (close) {
+        blockData = BlockData.newBuilder(blockData)
+            .addMetadata(KeyValue.newBuilder()
+                .setKey(OzoneConsts.LAST_PUT_FOR_BLOCK)
+                .setValue(Boolean.TRUE.toString()))
+            .build();
+      }
```

Comment:
I think it will bring in compatibility changes with the older client with no explicit close option set and latest server jar with new containers which will expect an explicit close option set? Are we ok with this?

Reply:
Good point, but that's only part of the problem.  Old client is anyway incompatible with new layout because it always sends 0 offset.  If backward compatibility is required, I think we would need to force old layout for old client.  For this, even block allocation needs to account for client version.

Reply:
I don't think we need to preserve protocol compatibility with old clients right now since we don't have rolling upgrades. As long as the old chunks are still readable after upgrade on DN we should be fine.